src/codegen/mod.rs

```diff
@@ -912,6 +912,40 @@ impl Expression {
                 Expression::Cast(self.loc(), to.clone(), Box::new(self.clone()))
             }
 
+            _ if !from.is_contract_storage()
+                && !to.is_contract_storage()
+                && from.is_reference_type(ns)
+                && !to.is_reference_type(ns) =>
+            {
+                let expr = Expression::Cast(
+                    self.loc(),
+                    Type::Uint(ns.target.ptr_size()),
+                    self.clone().into(),
+                );
+
+                expr.cast(to, ns)
+            }
+
+            _ if !from.is_contract_storage()
+                && !to.is_contract_storage()
+                && !from.is_reference_type(ns)
+                && to.is_reference_type(ns) =>
+            {
+                // cast non-pointer to pointer
+                let ptr_ty = Type::Uint(ns.target.ptr_size());
+
+                Expression::Cast(self.loc(), to.clone(), self.cast(&ptr_ty, ns).into())
+            }
+
+            _ if !from.is_contract_storage()
+                && !to.is_contract_storage()
+                && !from.is_reference_type(ns)
+                && !to.is_reference_type(ns) =>
+            {
+                // cast pointer to different pointer
+                Expression::Cast(self.loc(), to.clone(), self.clone().into())
+            }
```

Comment:
I understand this is the easiest solution to fix the bug, but it is complex to understand. 
Is it time for us to redesign `ast::Type` and perhaps include a pointer type (at another time, of course)?
I was thinking that we might also need a `codegen::Type`, because eliminating `Type::ExternalFunction` from codegen would clean up some messes there.

Reply:
I was thinking that when we introduce SSA, this becomes a new intermediate format.  So, CFG -> mid-level-IR and SSA becomes new low-level-IR.

Ideally the mid-level-IR does not have pointers (like our CFG now), but the low-level-IR does have pointers. So, I am not keen on introducing pointers in our CFG.